### PR TITLE
Improve hash parsing and errors

### PR DESCRIPTION
**Pull request by Ericson2314**
- Ensure hash is in form `<algo>-<prefix>` and not SRI.

 - Better errors if something goes wrong

 - `std::string_view` for no coppying

### CODE REVIEW

**Comment by Ericson2314** [comment]
WIP because will also fix some `libfetchers` FIXMEs.

**Comment by Ericson2314** [comment]
Hmm i cannot reproduce this test failure. It seems like the `echo -n '' > dummy` did not work.

**Comment by Ericson2314** [comment]
CC @domenkozar tests failed saying there's no `nix-build`. No what that's about?

File: src/libutil/parser.hh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Comment (edolstra) [inline src/libutil/parser.hh:1]: Maybe rename this file to `split.hh`? Since "parser" is pretty generic.

File: src/libutil/hash.cc [hunk 2]
```diff
@@ -136,63 +136,101 @@ std::string Hash::to_string(Base base, bool includeType) const
...
+Hash Hash::parseAnyPrefixed(std::string_view original)
+{
+    auto rest = original;
+    auto [optParsedType, isSRI] = getParsedTypeAndSRI(rest);
 
-    size_t size = s.size() - pos;
+    // Either the string or user must provide the type, if they both do they
+    // must agree.
+    if (!optParsedType)
+        throw BadHash("hash '%s' does not include a type.", rest);
```
Comment (edolstra) [inline src/libutil/hash.cc:180]: Style note: For consistency, no periods at the end of exception messages please.

**Comment by Ericson2314** [comment]
OK, fixed both.

**Comment by Ericson2314** [comment]
Fixed some more conflicts

**Comment by Ericson2314** [comment]
Ah, it is already fixed on this branch.

**Comment by Ericson2314** [comment]
Conflicts fixed.

**Comment by Ericson2314** [comment]
@edolstra wiggled the metadata so GitHub would recompute the diff, and this PR is actually readable.

**Comment by Ericson2314** [comment]
OK this should be ready.

**Pull request merged**